### DISCUSSION TITLE
Crash when deleting constrained body

### DISCUSSION THREAD

**Discussion by jankrassnigg** [jrouwe/JoltPhysics#519] [Q&A] [answered]
When I have two bodies joined with a constraint and then delete one of the bodies, Jolt crashes in ConstraintManager::GetActiveConstraints() because TwoBodyConstraint::IsValid() tries to access the body.

I think other physics engines handle this case transparently, ie the constraint just becomes useless, but nothing crashes. Do I need to keep track of which bodies are constrained myself, and make sure to delete the affected constraints as well? And if so, are there helper functions that could give me the list of constraints attached to a body?

**Comment by SirLynix** [comment]
I worked with multiple physics engine which did not handle this (Bullet, Chipmunk, Newton) and I wouldn't expect Jolt to do it. Unfortunately I think we should handle this ourselves.

**Comment by jrouwe** [comment, marked as answer]
Bodies don't keep track which constraints operate on them. This is by design because if they did, some sort of body locking would be needed when adding/removing constraints, which I'm trying to avoid for performance reasons. Also for performance reasons, constraints keep track of their attached bodies through pointers instead of body IDs. This means that if you delete a body but not the attached constraints then you have a dangling pointer and it will crash. In our case constraints are mostly used within a game object which also manages the bodies that the constraints connect to, so it's easy to remove constraints when the bodies are removed.

**Reply by jankrassnigg** [reply]
Thanks for confirming. I can implement the necessary book keeping on my side, just wanted to make sure that I didn't miss anything.
